fog/types/src/common.rs

```diff
@@ -61,6 +61,51 @@ impl BlockRange {
     pub fn is_empty(&self) -> bool {
         self.len() == 0
     }
+
+    /// Merge the first contiguous ranges into a range
+    ///
+    /// The ranges should be sorted to produce consistent results.
+    /// They are not sorted here in order to properly indicate the index of the
+    /// last range merged.
+    ///
+    /// ```
+    /// use mc_fog_types::common::BlockRange;
+    ///
+    /// let ranges = vec![BlockRange::new(0, 5), BlockRange::new(3, 8), BlockRange::new(8, 10), BlockRange::new(15, 16)];
+    /// let (index, merged_range) = BlockRange::merge_from_ranges(ranges.iter()).unwrap();
+    /// assert_eq!(index, 2);
+    /// assert_eq!(merged_range, BlockRange::new(0, 10));
+    ///
+    /// let ranges = vec![BlockRange::new(3, 8), BlockRange::new(0, 5), BlockRange::new(8, 10), BlockRange::new(15, 16)];
+    /// let (index, merged_range) = BlockRange::merge_from_ranges(ranges.iter()).unwrap();
+    /// assert_eq!(index, 0);
+    /// assert_eq!(merged_range, BlockRange::new(3, 8));
+    /// ```
+    ///
+    /// Returns the index in `ranges` of the last block range merged, and the
+    /// merged range
+    pub fn merge_from_ranges<'a>(
+        ranges: impl IntoIterator<Item = &'a BlockRange>,
+    ) -> Option<(usize, BlockRange)> {
+        let mut ranges = ranges.into_iter();
+        let first = ranges.next().cloned();
+        let Some(mut merged_range) = first else {
+            return None;
+        };
+        let mut index = 0;
+
+        for range in ranges {
+            if range.start_block <= merged_range.end_block
+                && merged_range.end_block <= range.end_block
+            {
+                merged_range.end_block = merged_range.end_block.max(range.end_block);
```

Comment:
small nit -- since this line is only reached when `merged_range.end_block` is less than or equal to `range.end_block` there is no need for a max operator.  I suppose the max could be left in case the enclosing logic changes in the future.
```suggestion
                merged_range.end_block = range.end_block;
```